[export] fix unbacked range deserialization

pianpwk opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/151809, by reading shape assertion nodes into ShapeEnv, and deferring instantiation of node example values, to be done node-by-node.

Differential Revision: D78588406

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -1876,33 +1899,23 @@ def deserialize_graph(self, serialized_graph: Graph) -> torch.fx.Graph:
-            meta_val = self.deserialize_tensor_meta(tensor_value)
-            log.debug("[deserialize_tensor_meta] %s (output): %s", name, meta_val)
-            self.serialized_name_to_meta[name] = meta_val
+            self.serialized_name_to_meta[name] = lambda v=tensor_value: self.deserialize_tensor_meta(v)
```
Have you checked Python semantics re: capture of `tensor_value` here? Hope it doesn't pick up values assigned later and instead does what you think it does...which is capture the value in *this* iteration?

pianpwk committed:
[export] fix unbacked range deserialization (#158681)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/151809, by reading shape assertion nodes into ShapeEnv, and deferring instantiation of node example values, to be done node-by-node.


Test Plan:
Imported from GitHub, without a `Test Plan:` line.

Rollback Plan:

Reviewed By: avikchaudhuri

Differential Revision: D78588406

Pulled By: pianpwk
